---
title: "Pathway Visualisation"
author: "Amartya Nambiar"
date: "2023-06-15"
categories: ["Notebook"]
description:  "&nbsp;Exploring the Pathway Visualisation Notebook by Alejandra Escobar (EBI Bioinformatician)"
title-block-banner: true
---
<div style="text-align:center">
![](pathway.png){width="500" height="300"}
</div>
<!-- #### What does it do? -->

#### Libraries Used

::: {.callout-note icon="false"}
# Importing Libraries

``` r
library(ALDEx2)
library(data.table)
library(dplyr)
library(IRdisplay)
library(KEGGREST)
library(MGnifyR)   
library(pathview)
library(tidyjson)
```

-   **ALDEx2** : Analyzing and comparing the abundance or occurrence of KEGG orthologs or modules within different samples or datasets.
-   **data.table**, **dplyr** : Used for data manipulation and transformation.
-   **KEGGREST** : Allows accessing KEGG pathways.
-   **MGnifyR** : It wraps the MGnify API in R functions.
-   **pathview** : Used for visualizing and analyzing pathway data with gene or protein expression data overlaid.
:::

#### Collecting Pathways using keggLink (KEGGREST Package)

::: {.callout-note icon="false"}
# Setting up Function for Pathways Collection

``` r
collect_pathways <- function(ids_list) {
    pathways = list()
    for (id in ids_list) { 
        current_pathway = as.list(keggLink("pathway", id))
        for (index in grep("map", current_pathway)) {        
            clean_id = gsub("*path:", "", current_pathway[index])
            # Discarding chemical structure (map010XX), global (map011XX), and overview (map012XX) maps
            prefix = substring(clean_id, 1, 6)
            if(is.na(match("map010", prefix)) & is.na(match("map011", prefix)) & is.na(match("map012", prefix)) ){
                pathways = append(pathways, clean_id)
            }
        }
    }
    return(pathways)
}
```

-   **Input** : List of Modules :: Output : List of Pathways
-   **keggLink()** is used to retrieve pathway info -\> *"pathway"*.
-   Using **grep** to get the *pathway maps* specifically.
-   Getting rid of some pathways<mark style="background-color:red">**TODO**</mark>
:::

#### MGnifyR Setup =\> Accession of Analyses

::: {.callout-note icon="false"}
# Sample Selection & Analyses Accession

``` r
mg = mgnify_client(usecache = T, cache_dir = '/home/jovyan/.mgnify_cache')
all_accessions = mgnify_analyses_from_studies(mg, c('MGYS00006180','MGYS00006178'))
all_metadata = mgnify_get_analyses_metadata(mg, all_accessions)
```

-   Both the studies chosen by [Alejandra Escobar](https://www.embl.org/people/person/alejandra-escobar-zepeda/) are related to the gut bacteria of honeybees.
:::
