<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Alejandra Escobar (MGnify team) and Amartya Nambiar">
<meta name="dcterms.date" content="2023-06-29">
<meta name="description" content="An expertly curated notebook to generate pathway visualizations from MGnify data, to which I actively contributed.">

<title>GSoC - 23 : The MGnify Project - Pathways Visualization</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../logo.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">GSoC - 23 : The MGnify Project</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://amartya.vercel.app/" rel="" target=""><i class="bi bi-globe" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/amartyanambiar" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/amartya-nambiar1/" rel="" target=""><i class="bi bi-linkedin" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../src/notebooks/PathwaysExplorer.html"><strong>Highlights</strong></a></li><li class="breadcrumb-item"><a href="../../src/notebooks/PathwaysExplorer.html">Pathways Visualization</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text"><strong>Highlights</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../src/notebooks/PathwaysExplorer.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Pathways Visualization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../posts/Viz_Recon/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">MGnify Visualization Recon</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text"><strong>About</strong></span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../The Project/project.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The Project</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../The Project/mentors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The Mentors</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../The Project/me.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Me</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Sidebar/keywords.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><strong>Keywords</strong></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../Sidebar/commands.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><strong>Saviour Commands</strong></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#pathways-visualization" id="toc-pathways-visualization" class="nav-link active" data-scroll-target="#pathways-visualization">Pathways Visualization</a>
  <ul class="collapse">
  <li><a href="#contents" id="toc-contents" class="nav-link" data-scroll-target="#contents">Contents</a></li>
  <li><a href="#drawing-presenceabsence-kos-for-one-metagenomic-sample" id="toc-drawing-presenceabsence-kos-for-one-metagenomic-sample" class="nav-link" data-scroll-target="#drawing-presenceabsence-kos-for-one-metagenomic-sample">Drawing presence/absence KOs for one metagenomic sample </a><a id="part1" class="nav-link" data-scroll-target="undefined"></a>
  <ul class="collapse">
  <li><a href="#fetching-data-from-mgnify-pathways-selection" id="toc-fetching-data-from-mgnify-pathways-selection" class="nav-link" data-scroll-target="#fetching-data-from-mgnify-pathways-selection">1.1. Fetching data from MGnify &amp; Pathways Selection </a><a id="part1_1" class="nav-link" data-scroll-target="undefined"></a></li>
  <li><a href="#selecting-the-most-complete-pathways" id="toc-selecting-the-most-complete-pathways" class="nav-link" data-scroll-target="#selecting-the-most-complete-pathways">1.2. Selecting the most complete pathways </a><a id="part1_2" class="nav-link" data-scroll-target="undefined"></a></li>
  <li><a href="#ready-to-draw" id="toc-ready-to-draw" class="nav-link" data-scroll-target="#ready-to-draw">1.3. Ready to draw! </a><a id="part1_3" class="nav-link" data-scroll-target="undefined"></a></li>
  </ul></li>
  <li><a href="#resources-and-documentation" id="toc-resources-and-documentation" class="nav-link" data-scroll-target="#resources-and-documentation">Resources and Documentation</a><a id="part2" class="nav-link" data-scroll-target="undefined"></a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Pathways Visualization</h1>
  <div class="quarto-categories">
    <div class="quarto-category">R</div>
    <div class="quarto-category">Visualization</div>
  </div>
  </div>

<div>
  <div class="description">
    An expertly curated notebook to generate pathway visualizations from MGnify data, to which I actively contributed.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Alejandra Escobar (MGnify team) and Amartya Nambiar </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">June 29, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<div style="max-width:1200px">
<img src="../_resources/mgnify_banner.png" width="100%">
</div>
<section id="pathways-visualization" class="level1">
<h1>Pathways Visualization</h1>
<p>In this notebook we aim to demonstrate how the MGnifyR tool can be used to fetch functional annotation results generated through the MGnify metagenomic analyisis pipelines. Then we show how to generate the pathways visualization using <a href="https://bioconductor.org/packages/release/bioc/html/pathview.html">Pathview</a> in R.</p>
<p><a href="http://github.com/beadyallen/mgnifyr">MGnifyR</a> is a library that provides a set of tools for easily accessing and processing MGnify data in R, making queries to MGnify databases through the <a href="https://www.ebi.ac.uk/metagenomics/api/v1/">MGnify API</a>. The benefit of MGnifyR is that data can either be fetched in tsv format or be directly combined in a phyloseq object to run an analysis in a custom workflow.</p>
<p>This is an interactive code notebook (a Jupyter Notebook). To run this code, click into each cell and press the ▶ button in the top toolbar, or press shift+enter</p>
<section id="contents" class="level2">
<h2 class="anchored" data-anchor-id="contents">Contents</h2>
<ul>
<li><a href="#part1">Drawing presence/absence KOs for one metagenomic sample</a>
<ul>
<li><a href="#part1_1">1.1. Fetching data from MGnify</a></li>
<li><a href="#part1_2">1.2. Selecting the most complete pathways</a></li>
<li><a href="#part1_3">1.3. Ready to draw!</a></li>
</ul></li>
<li><a href="#part2">Resources and Documentation</a></li>
</ul>
<p><strong>Libraries Used</strong></p>
<ul>
<li><strong>ALDEx2</strong> : Analyzing and comparing the abundance or occurrence of KEGG orthologs or modules within different samples or datasets.</li>
<li><strong>data.table</strong>, <strong>dplyr</strong> : Used for data manipulation and transformation.</li>
<li><strong>KEGGREST</strong> : Allows accessing KEGG pathways.</li>
<li><strong>MGnifyR</strong> : It wraps the MGnify API in R functions.</li>
<li><strong>pathview</strong> : Used for visualizing and analyzing pathway data with gene or protein expression data overlaid.</li>
<li><strong>variable_utils.r</strong> : Contains fucntion to retrieve variables from the environment &amp; user input.</li>
<li><strong>kegg_pathways_utils.r</strong> : Contains functions to fetch and process KEGG pathways.</li>
</ul>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loading libraries:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ALDEx2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(data.table)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(IRdisplay)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(KEGGREST)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(MGnifyR)   </span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pathview)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(tidyjson)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(IRdisplay)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"lib/variable_utils.r"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"lib/kegg_pathways_utils.r"</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">display_markdown</span>(<span class="at">file =</span> <span class="st">'../_resources/mgnifyr_help.md'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning message:
“package ‘MASS’ was built under R version 4.2.3”
Warning message:
“package ‘truncnorm’ was built under R version 4.2.3”
Warning message:
“package ‘dplyr’ was built under R version 4.2.3”</code></pre>
</div>
<section id="help-with-mgnifyr" class="level1 cell-output cell-output-display">
<h1>Help with MGnifyR</h1>
<p>MGnifyR is an R package that provides a convenient way for R users to access data from <a href="https://www.ebi.ac.uk/metagenomics/api/">the MGnify API</a>.</p>
<p>Detailed help for each function is available in R using the standard <code>?function_name</code> command (i.e.&nbsp;typing <code>?mgnify_query</code> will bring up built-in help for the mgnify_query command).</p>
<p>A vignette is available containing a reasonably verbose overview of the main functionality. This can be read either within R with the <code>vignette("MGnifyR")</code> command, or <a href="https://htmlpreview.github.io/?https://github.com/beadyallen/MGnifyR/blob/master/doc/MGnifyR.html">in the development repository</a></p>
<section id="mgnifyr-command-cheat-sheet" class="level2">
<h2 class="anchored" data-anchor-id="mgnifyr-command-cheat-sheet">MGnifyR Command cheat sheet</h2>
<p>The following list of key functions should give a starting point for finding relevent documentation.</p>
<ul>
<li><code>mgnify_client()</code> : Create the client object required for all other functions.</li>
<li><code>mgnify_query()</code> : Search the whole MGnify database.</li>
<li><code>mgnify_analyses_from_xxx()</code> : Convert xxx accessions to analyses accessions. xxx is either samples or studies.</li>
<li><code>mgnify_get_analyses_metadata()</code> : Retrieve all study, sample and analysis metadata for given analyses.</li>
<li><code>mgnify_get_analyses_phyloseq()</code> : Convert abundance, taxonomic, and sample metadata into a single phyloseq object.</li>
<li><code>mgnify_get_analyses_results()</code> : Get functional annotation results for a set of analyses.</li>
<li><code>mgnify_download()</code> : Download raw results files from MGnify.</li>
<li><code>mgnify_retrieve_json()</code> : Low level API access helper function.</li>
</ul>
</section>
</section>
</div>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="2">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create your session mgnify_client object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>mg <span class="ot">=</span> <span class="fu">mgnify_client</span>(<span class="at">usecache =</span> T, <span class="at">cache_dir =</span> <span class="st">'/home/jovyan/.mgnify_cache'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Setting tables and figures size to display (these will be reset later):</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.matrix.max.cols=</span><span class="dv">150</span>, <span class="at">repr.matrix.max.rows=</span><span class="dv">500</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">repr.plot.width=</span><span class="dv">4</span>, <span class="at">repr.plot.height=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="drawing-presenceabsence-kos-for-one-metagenomic-sample" class="level2">
<h2 class="anchored" data-anchor-id="drawing-presenceabsence-kos-for-one-metagenomic-sample">Drawing presence/absence KOs for one metagenomic sample <a id="part1"></a></h2><a id="part1">
</a><section id="fetching-data-from-mgnify-pathways-selection" class="level3"><a id="part1">
</a><h3 class="anchored" data-anchor-id="fetching-data-from-mgnify-pathways-selection"><a id="part1">1.1. Fetching data from MGnify &amp; Pathways Selection </a><a id="part1_1"></a></h3><a id="part1_1">
</a><div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="4"><a id="part1_1">
</a><div class="sourceCode cell-code" id="cb5"><a id="part1_1"></a><pre class="sourceCode r code-with-copy"><a id="part1_1"><code class="sourceCode r"><span id="cb5-1"></span></code></a><code class="sourceCode r"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>PATHWAY_STUDY_IDS <span class="ot">&lt;-</span> <span class="fu">get_variable_from_link_or_input</span>(<span class="st">'PATHWAY_STUDY_IDS'</span>, <span class="at">name =</span>  <span class="st">'Study Accession'</span>, <span class="at">default =</span> <span class="st">'MGYS00006180,MGYS00006178'</span>)
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>PATHWAY_STUDY_IDS <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">strsplit</span>(PATHWAY_STUDY_IDS, <span class="st">","</span>)[[<span class="dv">1</span>]])</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Using"</span>, PATHWAY_STUDY_IDS, <span class="st">"as Study Accession"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Using Study Accession = MGYS00006180,MGYS00006178 from the link you followed.
[1] "Using MGYS00006180 as Study Accession"
[2] "Using MGYS00006178 as Study Accession"</code></pre>
</div>
</div>
<ul>
<li>Both the studies chosen by <a href="https://www.embl.org/people/person/alejandra-escobar-zepeda/">Alejandra Escobar</a> are related to the gut bacteria of honeybees.</li>
</ul>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Custom Pathways Selection</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Most general pathways include</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Glycolysis / Gluconeogenesis, Citrate cycle (TCA cycle), Pentose phosphate pathway, </span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Fatty acid biosynthesis, Pyrimidine metabolism, Oxidative phosphorylation</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>CUSTOM_PATHWAY_IDS <span class="ot">&lt;-</span> <span class="fu">PathwaysSelection</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="pathways-selection" class="level4 cell-output cell-output-display">
<h4 class="anchored" data-anchor-id="pathways-selection">Pathways Selection :</h4>
<ul>
<li><p>For the most general &amp; most complete pathways, input ‘G’</p></li>
<li><p>Press Enter to generate the most complete pathways</p></li>
<li><p>To add custom pathways, input pathway numbers (ex: 00053,01220)</p></li>
</ul>
</section>
<div class="cell-output cell-output-stdout">
<pre><code>Using Pathways Accession = G from the link you followed.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Using 00010  -  Glycolysis / Gluconeogenesis  :  https://www.kegg.jp/pathway/map00010 as a Custom Pathway
Using 00020  -  Citrate cycle (TCA cycle)  :  https://www.kegg.jp/pathway/map00020 as a Custom Pathway
Using 00030  -  Pentose phosphate pathway  :  https://www.kegg.jp/pathway/map00030 as a Custom Pathway
Using 00061  -  Fatty acid biosynthesis  :  https://www.kegg.jp/pathway/map00061 as a Custom Pathway
Using 01232  -  Nucleotide metabolism  :  https://www.kegg.jp/pathway/map01232 as a Custom Pathway
Using 00240  -  Pyrimidine metabolism  :  https://www.kegg.jp/pathway/map00240 as a Custom Pathway
Using 00190  -  Oxidative phosphorylation  :  https://www.kegg.jp/pathway/map00190 as a Custom Pathway
</code></pre>
</div>
</div>
<ol type="1">
<li>Fetching the analysis accession list using the study accessions.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>output <span class="ot">&lt;-</span> <span class="fu">capture.output</span>({</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  all_accessions <span class="ot">&lt;-</span> <span class="fu">mgnify_analyses_from_studies</span>(mg, PATHWAY_STUDY_IDS)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  all_metadata <span class="ot">&lt;-</span> <span class="fu">mgnify_get_analyses_metadata</span>(mg, all_accessions)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Keeping just the first analysis accession to fetch the kegg orthologs count table from the MGnify API and transform from JSON to matrix.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>accession <span class="ot">=</span> <span class="fu">head</span>(all_accessions, <span class="dv">1</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>ko_loc <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">'analyses/'</span>,accession,<span class="st">'/kegg-orthologs'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>ko_json <span class="ot">=</span> <span class="fu">mgnify_retrieve_json</span>(mg, <span class="at">path =</span> ko_loc)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>ko_data <span class="ot">=</span> <span class="fu">as.data.frame</span>(ko_json <span class="sc">%&gt;%</span> spread_all)[ , <span class="fu">c</span>(<span class="st">"attributes.accession"</span>, <span class="st">"attributes.count"</span>)]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>ko_data <span class="ot">=</span> <span class="fu">data.frame</span>(ko_data, <span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ko_data)[<span class="dv">1</span>] <span class="ot">=</span> <span class="st">'counts'</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>ko_matrix <span class="ot">=</span> <span class="fu">data.matrix</span>(ko_data)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ko_matrix, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true">
<caption>A matrix: 3 × 1 of type dbl</caption>
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th" scope="col">counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">K10822</td>
<td>115</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th" scope="row">K02030</td>
<td>91</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th" scope="row">K02004</td>
<td>79</td>
</tr>
</tbody>
</table>
</div>
</div>
<ol start="3" type="1">
<li>Fetch the modules completeness table and filter out completeness &lt; 100%.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>comp_loc <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">'analyses/'</span>,accession,<span class="st">'/kegg-modules'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>ko_comp_json <span class="ot">=</span> <span class="fu">mgnify_retrieve_json</span>(mg, <span class="at">path =</span> comp_loc)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>ko_comp <span class="ot">=</span> <span class="fu">as.data.frame</span>(ko_comp_json <span class="sc">%&gt;%</span> spread_all)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>modules <span class="ot">=</span> ko_comp[ko_comp<span class="sc">$</span>attributes.completeness <span class="sc">==</span> <span class="dv">100</span>,][, <span class="fu">c</span>(<span class="st">"attributes.accession"</span>)]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(modules)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<style>
.list-inline {list-style: none; margin:0; padding: 0}
.list-inline>li {display: inline-block}
.list-inline>li:not(:last-child)::after {content: "\00b7"; padding: 0 .5ex}
</style>
<ol class="list-inline"><li>'M00001'</li><li>'M00002'</li><li>'M00003'</li><li>'M00004'</li><li>'M00005'</li><li>'M00006'</li></ol>
</div>
</div>
<hr>
</section>
<section id="selecting-the-most-complete-pathways" class="level3">
<h3 class="anchored" data-anchor-id="selecting-the-most-complete-pathways">1.2. Selecting the most complete pathways <a id="part1_2"></a></h3><a id="part1_2">
</a><div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="10"><a id="part1_2">
</a><div class="sourceCode cell-code" id="cb14"><a id="part1_2"></a><pre class="sourceCode r code-with-copy"><a id="part1_2"><code class="sourceCode r"><span id="cb14-1"></span></code></a><code class="sourceCode r"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Setting up function that collects KEGG pathways for a given list of IDs, excluding chemical structure &amp; global maps</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>collect_pathways <span class="ot">&lt;-</span> <span class="cf">function</span>(ids_list) {</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    pathways <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (id <span class="cf">in</span> ids_list) { </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>        current_pathway <span class="ot">=</span> <span class="fu">as.list</span>(<span class="fu">keggLink</span>(<span class="st">"pathway"</span>, id))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (index <span class="cf">in</span> <span class="fu">grep</span>(<span class="st">"map"</span>, current_pathway)) {        </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>            clean_id <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">"*path:"</span>, <span class="st">""</span>, current_pathway[index])</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Discarding chemical structure (map010XX), global (map011XX), and overview (map012XX) maps</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>            prefix <span class="ot">=</span> <span class="fu">substring</span>(clean_id, <span class="dv">1</span>, <span class="dv">6</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span>(<span class="fu">is.na</span>(<span class="fu">match</span>(<span class="st">"map010"</span>, prefix)) <span class="sc">&amp;</span> <span class="fu">is.na</span>(<span class="fu">match</span>(<span class="st">"map011"</span>, prefix)) <span class="sc">&amp;</span> <span class="fu">is.na</span>(<span class="fu">match</span>(<span class="st">"map012"</span>, prefix)) ){</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>                pathways <span class="ot">=</span> <span class="fu">append</span>(pathways, clean_id)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(pathways)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><strong>Input</strong> : List of Modules :: Output : List of Pathways</li>
<li><strong>keggLink()</strong> is used to retrieve pathway info -&gt; <em>“pathway”</em>.</li>
<li>Using <strong>grep</strong> to get the <em>pathway maps</em> specifically.</li>
</ul>
<ol type="1">
<li>Now we need to collect the list of template pathways where these complete modules can be draw. This step takes less than 1 minute to run.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>md_pathways <span class="ot">=</span> <span class="fu">collect_pathways</span>(modules)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(md_pathways)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<ol>
    <li>'map00010'</li>
    <li>'map00010'</li>
    <li>'map00010'</li>
    <li>'map00020'</li>
    <li>'map00030'</li>
    <li>'map00030'</li>
</ol>
</div>
</div>
<ol start="2" type="1">
<li>In order to draw the most complete pathways maps, we will use the list of templates we obtained in the previous step and select only pathways having all their constituent modules.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Counting the number of modules we have in each pathway</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>our_pathways_counts <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (path_element <span class="cf">in</span> md_pathways) {</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (path_element <span class="sc">%in%</span> <span class="fu">names</span>(our_pathways_counts)) {</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>        new_value <span class="ot">=</span> our_pathways_counts[[path_element]] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>        our_pathways_counts[path_element] <span class="ot">=</span> new_value       </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>        our_pathways_counts[path_element] <span class="ot">=</span> <span class="dv">1</span> </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Counting the number of modules expected in each pathway</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>u_pathways <span class="ot">=</span> <span class="fu">unique</span>(md_pathways)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>exp_pathways_counts <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (path <span class="cf">in</span> u_pathways) {</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    mod_count <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">as.list</span>(<span class="fu">keggLink</span>(<span class="st">"module"</span>, path)))</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    exp_pathways_counts[path] <span class="ot">=</span> mod_count </span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting the pathways having all their constituent modules. We remove the 'map' prefix as pathview doesn't like it</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>to_draw <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (pathway <span class="cf">in</span> <span class="fu">names</span>(our_pathways_counts)) {</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>    our_value <span class="ot">=</span> our_pathways_counts[[pathway]]</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>    exp_value <span class="ot">=</span> exp_pathways_counts[[pathway]]</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>    ratio <span class="ot">=</span>  our_value <span class="sc">/</span> exp_value</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (ratio <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>        nude_id <span class="ot">=</span>  <span class="fu">gsub</span>(<span class="st">"map"</span>, <span class="st">""</span>, pathway)</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>        to_draw <span class="ot">=</span> <span class="fu">append</span>(to_draw, nude_id)   </span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding the custom pathways to to_draw if not present already</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (pathway <span class="cf">in</span> CUSTOM_PATHWAY_IDS){</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span>(pathway <span class="sc">%in%</span> to_draw)) {</span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>    to_draw <span class="ot">=</span> <span class="fu">append</span>(to_draw, pathway)</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>    }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># printing name of the pathways to be drawn</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (pathway <span class="cf">in</span> to_draw){</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">display_markdown</span>(<span class="fu">paste</span>(pathway, <span class="st">"--&gt; "</span>, <span class="fu">get_pathway_info</span>(pathway)[<span class="dv">1</span>],<span class="st">" : "</span>,<span class="fu">get_pathway_info</span>(pathway)[<span class="dv">2</span>], <span class="at">sep=</span><span class="st">" "</span>))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p>00010 –&gt; Glycolysis / Gluconeogenesis : https://www.kegg.jp/pathway/map00010</p>
</div>
<div class="cell-output cell-output-display">
<p>00480 –&gt; Glutathione metabolism : https://www.kegg.jp/pathway/map00480</p>
</div>
<div class="cell-output cell-output-display">
<p>00430 –&gt; Taurine and hypotaurine metabolism : https://www.kegg.jp/pathway/map00430</p>
</div>
<div class="cell-output cell-output-display">
<p>00521 –&gt; Streptomycin biosynthesis : https://www.kegg.jp/pathway/map00521</p>
</div>
<div class="cell-output cell-output-display">
<p>00020 –&gt; Citrate cycle (TCA cycle) : https://www.kegg.jp/pathway/map00020</p>
</div>
<div class="cell-output cell-output-display">
<p>00030 –&gt; Pentose phosphate pathway : https://www.kegg.jp/pathway/map00030</p>
</div>
<div class="cell-output cell-output-display">
<p>00061 –&gt; Fatty acid biosynthesis : https://www.kegg.jp/pathway/map00061</p>
</div>
<div class="cell-output cell-output-display">
<p>01232 –&gt; Nucleotide metabolism : https://www.kegg.jp/pathway/map01232</p>
</div>
<div class="cell-output cell-output-display">
<p>00240 –&gt; Pyrimidine metabolism : https://www.kegg.jp/pathway/map00240</p>
</div>
<div class="cell-output cell-output-display">
<p>00190 –&gt; Oxidative phosphorylation : https://www.kegg.jp/pathway/map00190</p>
</div>
</div>
<hr>
</section>
<section id="ready-to-draw" class="level3">
<h3 class="anchored" data-anchor-id="ready-to-draw">1.3. Ready to draw! <a id="part1_3"></a></h3><a id="part1_3">
<ol type="1">
<li>As we are plotting absence/presence, we set the number of bins = 2, the scale in one direction, and use 1 as limit.</li>
</ol>
<p>Takes a couple of minutes depending on the number of pathways</p>
</a><div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="14"><a id="part1_3">
</a><div class="sourceCode cell-code" id="cb18"><a id="part1_3"></a><pre class="sourceCode r code-with-copy"><a id="part1_3"><code class="sourceCode r"><span id="cb18-1"></span></code></a><code class="sourceCode r"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>({
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (p <span class="cf">in</span> to_draw) {</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pathview</span>(<span class="at">gene.data =</span> ko_matrix, </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">species =</span> <span class="st">"ko"</span>, </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">pathway.id =</span> p, </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">bins=</span><span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>), </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">both.dirs =</span> <span class="cn">FALSE</span>, </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">limit =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">mid =</span> <span class="fu">c</span>(<span class="st">"#ffffff"</span> , <span class="st">"#ffffff"</span>), </span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">high =</span> <span class="fu">c</span>(<span class="st">"#02b3ad"</span> , <span class="st">"#02b3ad"</span>)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Clearing the current working directory and Displaying all the generated figures that are stored at the <code>pathway_plots/</code> directory.</li>
</ol>
<div class="cell" data-tags="[]" data-vscode="{&quot;languageId&quot;:&quot;r&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">generatePathwayPlots</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p>Glycolysis / Gluconeogenesis</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Citrate cycle (TCA cycle)</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-4.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Pentose phosphate pathway</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-6.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Fatty acid biosynthesis</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-8.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Oxidative phosphorylation</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-10.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Pyrimidine metabolism</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-12.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Taurine and hypotaurine metabolism</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-14.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Glutathione metabolism</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-16.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Streptomycin biosynthesis</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-18.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p>Nucleotide metabolism</p>
</div>
<div class="cell-output cell-output-display">
<p><img src="PathwaysExplorer_files/figure-html/cell-16-output-20.png" class="img-fluid"></p>
</div>
</div>
<hr>
</section>
</section>
<section id="resources-and-documentation" class="level2">
<h2 class="anchored" data-anchor-id="resources-and-documentation">Resources and Documentation<a id="part2"></a></h2><a id="part2">
</a><section id="pathview" class="level4"><a id="part2">
<h4 class="anchored" data-anchor-id="pathview"><strong>Pathview </strong>:</h4>
</a><ul><a id="part2">
<li>The official documentation for Pathview, a tool for pathway-based data integration and visualization. It provides an overview of the tool and explains how to use it effectively.</li>
</a><li><a id="part2">You can find more information at </a><a href="https://pathview.uncc.edu/overview#kegg_view">https://pathview.uncc.edu/overview#kegg_view</a> .</li>
</ul>
</section>
<section id="keggrest" class="level4">
<h4 class="anchored" data-anchor-id="keggrest"><strong>KEGGREST</strong>:</h4>
<ul>
<li>KEGGREST is a Bioconductor package that provides programmatic access to the KEGG database.</li>
<li>More about KEGGREST at <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/KEGGREST/inst/doc/KEGGREST-vignette.html">http://www.bioconductor.org/packages/release/bioc/vignettes/KEGGREST/inst/doc/KEGGREST-vignette.html</a> .</li>
</ul>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>