---
title: "MGnify Visualization Recon"
author: "Amartya Nambiar"
date: "2023-06-12"
categories: [Visualization]
description:  "Condensed list of Visualizations used in MGnify"
---


<div style="width:70%;">

![](mgnify.png)

</div>



# Sample Viz
## Map of sample locations

**Example URL:**
[[https://www.ebi.ac.uk/metagenomics/studies/MGYS00001606#overview]{.underline}](https://www.ebi.ac.uk/metagenomics/studies/MGYS00001606#overview)

**What it shows:** Location of the samples collected

**Dataset:** Sample metadata (from ENA)

**Visualization type:** Geographical Map

**Technology used:** Google Maps embed



# Analysis Viz

## Quality Control

**Example URL:**
[[https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#qc]{.underline}](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#qc)

**Technology used:** Highcharts

**Dataset:** Accessed through Analyses Accession ID - Summary

#### <u>Number of Sequence Reads per QC step</u> 

**What it shows:** Number of reads which pass the quality control steps
in the pipeline.

**Visualization type:** Normal X-axis Charts

#### <u>Read Length & Read GC Distribution</u>

**What it shows:** Histograms showing distributions of sequence lengths
and percentage GC content for the sequences having passed quality
control.

**Visualization type:** Histogram

#### <u>Nucleotide Position Histogram</u>

**What it shows:** Shows the relative abundance of nucleotides (A, C, G,
T, or ambiguous base \"N\") at each position starting from the beginning
of each read up to the first 500 base pairs.

**Visualization type:** Histogram

**Dataset:** Accessed through Analyses Accession ID -
nucleotide-distribution

## Taxonomic analysis

Results from the taxonomic analysis steps of the pipeline.

**Example URL:**
[[https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#taxonomic]{.underline}](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#taxonomic)

#### <u>Krona</u>

**What it shows:** Interactive Krona to visualise the taxa hierarchy

**Technology used:**
[[https://github.com/marbl/Krona/wiki]{.underline}](https://github.com/marbl/Krona/wiki)

**Visualization type:** Krona

**Dataset:** Accessed through Analyses Accession ID -
krona\${taxResults}

#### <u>Domain Composition</u>

**Technology used:** Highcharts

**Visualization type:** Pie, Column

**Dataset:** Accessed through Analysis Accession ID - taxonomy

#### <u>Phylum Composition</u>

**Technology used:** Highcharts

**Visualization type:** Pie, Column, Stacked Column

**Dataset:** Accessed through Analysis Accession ID - taxonomy

## Functional analysis

Charts present the functional analysis outputs of the pipeline, which
focus on InterPro, Pfam, KEGG orthologue and GO term annotations. These
summarise the functional content of the sequences in the sample.

**Example URL:**
[[https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#functional]{.underline}](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00584637#functional)

#### <u>InterPro</u>

**What it shows:** Shows Sequence feature summary & InterPro Summary

**Technology used:** Highcharts

**Visualization type:** X-axis Chart, Pie

#### <u>GO Terms</u>

**What it shows:** A summary of Gene Ontology (GO) terms derived from
InterPro matches to the sample is provided in the charts below.

**Technology used:** Highcharts

**Visualization type:** Bar, Pie

#### <u>Pfam</u>

**What it shows:** Top 10 Pfam entries

**Technology used:** Highcharts

**Visualization type:** Bar

#### <u>KO</u>

**What it shows:** Top 10 KO entries

**Technology used:** Highcharts

**Visualization type:** Bar

## Contig Viewer

**Example URL:**
[[https://www.ebi.ac.uk/metagenomics/analyses/MGYA00643204?selected_contig=ERZ12633542.2-NODE-96-length-2072-cov-5.941001#contigs-viewer]{.underline}](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00643204?selected_contig=ERZ12633542.2-NODE-96-length-2072-cov-5.941001#contigs-viewer)

**What it shows:** Contig browser for the assembly contigs

**Dataset:** through assembly ID, FASTA URL

**Visualization type:** Contig viewer

**Technology used:** IGV

## Pathways/Systems

Results from the biochemical pathways and systems predictions steps of
the pipeline.

**Example URL:** [https://www.ebi.ac.uk/metagenomics/analyses/MGYA00572069#path-systems](https://www.ebi.ac.uk/metagenomics/analyses/MGYA00572069#path-systems)

#### <u>KEGG Module</u>
**What it shows:** Modules vs Completeness

**Dataset:** `analyses/${overviewData.id}/kegg-modules`

**Visualization type:** Bar Chart

**Technology used:** Highcharts

#### <u>Genome properties</u>
**What it shows:** Hierarchical structure of Genome Properties

**Dataset:** `analyses/${overviewData.id}/genome-properties`

**Visualization type:** Hierarchy tree

**Technology used:** [HierarchyNode](https://github.com/EBI-Metagenomics/ebi-metagenomics-client/blob/db0402d572863358a9aab20136a2e9713a31fb2d/src/components/UI/Hierarchy/index.tsx)

#### <u>antiSMASH</u>
**What it shows:** Displays top 10 antiSMASH gene clusters

**Dataset:** `analyses/${overviewData.id}/antismash-gene-clusters`

**Visualization type:** Bar Chart

**Technology used:** Highcharts


# Genome Viz
**Example URL:** [https://www.ebi.ac.uk/metagenomics/genomes/MGYG000320984#genome-browser](https://www.ebi.ac.uk/metagenomics/genomes/MGYG000320984#genome-browser)

#### <u>Browse genome</u>
**What it shows:** A container with IGV genome browser, also allows selection of feature types

**Dataset:** General feature format : `${accession}.gff`, fasta : `${accession}.fna`

**Visualization type:** Genome browser + Functional Annotation

**Technology used:** IGV


#### <u>COG Analysis</u>
**What it shows:** Displays Top 10 COG Categories

**Dataset:** `${accession}/cogs`

**Visualization type:** Bar Chart

**Technology used:** Highcharts


#### <u>KEGG Class Analysis</u>
**What it shows:** Displays Top 10 KEGG brite Categories

**Dataset:** `${accession}/kegg-class`

**Visualization type:** Bar Chart

**Technology used:** Highcharts


#### <u>KEGG Module Analysis</u>
**What it shows:** Displays Top 10 KEGG module Categories

**Dataset:** `${accession}/kegg-module`

**Visualization type:** Bar Chart

**Technology used:** Highcharts


#### <u>Taxonomy Tree</u>
**Example URL:** [https://www.ebi.ac.uk/metagenomics/genome-catalogues/chicken-gut-v1-0#phylo-tab](https://www.ebi.ac.uk/metagenomics/genome-catalogues/chicken-gut-v1-0#phylo-tab)

**What it shows:** Phylogenetic Tree

**Dataset:** phylo_tree.json

**Visualization type:** [HierarchyNode](https://github.com/EBI-Metagenomics/ebi-metagenomics-client/blob/db0402d572863358a9aab20136a2e9713a31fb2d/src/components/UI/Hierarchy/index.tsx) 

---

# Kegg pathway Visualization

**Example URL:**
[[https://github.com/EBI-Metagenomics/notebooks/blob/pathways_vis/src/notebooks/R%20Examples/pathways_vis.ipynb]{.underline}](https://github.com/EBI-Metagenomics/notebooks/blob/pathways_vis/src/notebooks/R%20Examples/pathways_vis.ipynb)
(not yet live)

**What it shows:** completeness of KEGG pathways for a sample

**Dataset:** combination of an MGYA Kegg Module list and a Pathway from
KEGG API

**Visualization type:** pathway network

**Technology used:** pathview R package

<!-- ![](mgnify.png) -->